server/src/com/thoughtworks/go/config/update/AgentsEntityConfigUpdateCommand.java

```diff
@@ -0,0 +1,155 @@
+    private final TriState enable;
+    private final List<String> resourcesToAdd;
+    private final List<String> resourcesToRemove;
+    private GoConfigService goConfigService;
+    public Agents agents;
+
+    public AgentsEntityConfigUpdateCommand(Username username, HttpOperationResult result, List<String> uuids, List<String> environmentsToAdd, List<String> environmentsToRemove, TriState enable, List<String> resourcesToAdd, List<String> resourcesToRemove, GoConfigService goConfigService) {
+        this.username = username;
+        this.result = result;
+        this.uuids = uuids;
+        this.environmentsToAdd = environmentsToAdd;
+        this.environmentsToRemove = environmentsToRemove;
+        this.enable = enable;
+        this.resourcesToAdd = resourcesToAdd;
+        this.resourcesToRemove = resourcesToRemove;
+        this.goConfigService = goConfigService;
+    }
+
+    @Override
+    public boolean canContinue(CruiseConfig cruiseConfig) {
+        if (goConfigService.isAdministrator(username.getUsername().toString())) {
+            return true;
+        }
+
+        result.unauthorized("Unauthorized to operate on agent", "Unauthorized to operate on agent", HealthStateType.general(HealthStateScope.GLOBAL));
+        return false;
+    }
+
+    private List<AgentConfig> getValidAgents(List<String> uuids, HttpOperationResult result, Agents agents) throws NoSuchAgentException {
+        List<AgentConfig> goodAgents = new ArrayList<>();
+        List<String> unknownAgents = new ArrayList<>();
+
+        for (String uuid : uuids) {
+            AgentConfig agent = agents.getAgentByUuid(uuid);
+            if (!agent.isNull()) {
+                goodAgents.add(agent);
+            } else {
+                unknownAgents.add(uuid);
+            }
+        }
+        if (!unknownAgents.isEmpty()) {
+            NoSuchAgentException noAgentException = new NoSuchAgentException(unknownAgents);
+            result.unprocessibleEntity("NO_SUCH_AGENTS", noAgentException.getMessage(), HealthStateType.general(HealthStateScope.GLOBAL));
+            throw noAgentException;
+        }
+
+        return goodAgents;
+    }
+
+    private void validateEnvironment(Set<CaseInsensitiveString> allEnvironmentNames, List<String> environmentsToOperate, HttpOperationResult result) throws NoSuchEnvironmentException {
+        for (String environment : environmentsToOperate) {
+            CaseInsensitiveString environmentName = new CaseInsensitiveString(environment);
+            if (!allEnvironmentNames.contains(environmentName)) {
+                NoSuchEnvironmentException exception = new NoSuchEnvironmentException(environmentName);
+                result.unprocessibleEntity("NO_SUCH_ENVIRONMENT", exception.getMessage(), HealthStateType.general(HealthStateScope.GLOBAL));
+                throw exception;
+            }
+        }
+    }
+
+    @Override
+    public void update(CruiseConfig preprocessedConfig) throws Exception {
```

Comment:
This can be condensed to 
```
        List<AgentConfig> goodAgents = getValidAgents(uuids, result, preprocessedConfig.agents());

        Set<CaseInsensitiveString> allEnvironmentNames = new HashSet<>(goConfigService.getEnvironments().names());

        validateEnvironment(allEnvironmentNames, environmentsToAdd, result);
        validateEnvironment(allEnvironmentNames, environmentsToRemove, result);

        for (AgentConfig agentConfig : goodAgents) {
            agentConfig.disable(enable.isFalsy());

            for (String r : resourcesToAdd) {
                agentConfig.addResource(new Resource(r));
            }

            for (String r : resourcesToRemove) {
                agentConfig.removeResource(new Resource(r));
            }

            for (String environment : environmentsToAdd) {
                EnvironmentConfig environmentConfig = preprocessedConfig.getEnvironments().find(new CaseInsensitiveString(environment));
                if (environmentConfig != null && !environmentConfig.hasAgent(agentConfig.getUuid()))
                    environmentConfig.addAgent(agentConfig.getUuid());
            }

            for (String environment : environmentsToRemove) {
                EnvironmentConfig environmentConfig = preprocessedConfig.getEnvironments().find(new CaseInsensitiveString(environment));
                if (environmentConfig != null && !environmentConfig.hasAgent(agentConfig.getUuid()))
                    environmentConfig.removeAgent(agentConfig.getUuid());
            }

        }
```